File: src/nodejs/websocket.js
const { Socket } = require('phoenix-channels')

let socket = new Socket("wss://ocap.arcblock.io/api/socket");

socket.connect();

// Now that you are connected, you can join channels with a topic:
let channel = socket.channel('__absinthe__:control');

const query = "subscription { \n  newBlockMined { \n    hash\n }\n}";

channel.join()
  .receive("ok", resp => {
    console.log("Joined successfully", resp)
    channel.push("doc", { query: query })
      .receive("ok", resp => {
        console.log(resp);
        channel.on(resp, resp => {
          console.log(resp);
        })
      })
  })
  .receive("error", resp => { console.log("Unable to join", resp) })

socket.onMessage(({ event, payload, ref }) => {
  if (event === "subscription:data") {
    console.log(payload.result.data);
  }

});

File: src/python/websocket.py
from chunnel import Socket

query = "subscription { \n  newBlockMined { \n    hash\n }\n}"

async def listen():
  socket = Socket('wss://ocap.arcblock.io/api/socket/websocket', {})
  async with socket:
    channel = socket.channel('__absinthe__:control', {})
    response = await channel.join()
    print("Joined:", response)
    response = await channel.push('doc', {'query': query})
    print("Topic listened:", await response.response())
    message = await channel.receive()
    data = message.payload['result']['data']
    print(data)

def main():
  import asyncio
  loop = asyncio.get_event_loop()
  loop.run_until_complete(listen())
  loop.run_forever()

if __name__ == '__main__':
  main()


# 16:23:35.436 application=absinthe module=Absinthe.Subscription.Local function=run_docset/2 [debug] AbsintheSubscription Publication
# Field Topic: {:new_block_mined, :new_block}
# Subscription id: "__absinthe__:doc:80613059"
# Data: %{data: %{"newBlockMined" => %{"hash" => "0xf577a1973cbd7cf3936e4ab4c289d9d3d3cfff7e0e29d55ad8b7e64fc88c38f7"}}}
